### ISSUE TITLE
AutoCatBoostCARMA doesn't forecast

### ISSUE THREAD

**Issue by joclimber** [AdrianAntico/RemixAutoML#76]
Hi, I'm trying to forecast a single time series using AutoCatBoostCARMA. When the fuction finishes processing I can't find the forecast. I'm pretty sure I'm using it wrong, but can't figure out why.  I want to forecast from 2018-12-01 to 2020-01-01 (13 months).

My CatBoostResults object contains: Model, TrainData, TestData, PlotList, EvaluationMetrics, VariableImportance, InteractionImportance, GridMetrics, ColNames, TransformationResults, FactorLevelsList. Can't find the forecast in there. 

Grateful for any tips I can get :)

Here is the code i'm running:

        # creating data

        data <- data.frame(month=c('2016-01-01','2016-02-01','2016-03-01','2016-04-01','2016-05-01','2016-06-01','2016-07-01','2016-08-01','2016-09-01','2016-10-01','2016-11-01','2016-12-01','2017-01-01','2017-02-01','2017-03-01','2017-04-01','2017-05-01','2017-06-01','2017-07-01','2017-08-01','2017-09-01','2017-10-01','2017-11-01','2017-12-01','2018-01-01','2018-02-01','2018-03-01','2018-04-01','2018-05-01','2018-06-01','2018-07-01','2018-08-01','2018-09-01','2018-10-01','2018-11-01'), value = c(20,280,200,90,100,40,70,210,90,0,10,70,10,90,300,0,70,40,90,100,60,80,40,140,30,130,70,110,60,20,10,220,330,1,134)) %>%
                dplyr::mutate(month = zoo::as.Date(month)

        # running function

        TimeUnit = "months"

        CatBoostResults <- RemixAutoML::AutoCatBoostCARMA(
        
            # data args
            data = data ,
            TargetColumnName = "month",
            DateColumnName = "value ",
            HierarchGroups = NULL,
            TimeUnit = TimeUnit,
            TimeGroups = if(TimeUnit == "weeks") c("weeks","months","quarters") else if(TimeUnit == "months") c("months","quarters"),
            
            # Production args
            TrainOnFull = FALSE,
            PartitionType = "timeseries", 
            FC_Periods = 13
            TaskType = "CPU",
            Timer = TRUE,
            DebugMode = TRUE,
            
            # Target variable transformations
            TargetTransformation = FALSE,
            Methods = c("BoxCox","Asinh","Log","LogPlus1","YeoJohnson"),
            Difference = FALSE,
            NonNegativePred = TRUE,
            RoundPreds = FALSE,
            
            # Calendar-related features
            CalendarVariables = if(TimeUnit == "weeks") c("week","wom","month","quarter") else if(TimeUnit == "months") c("month","quarter"), 
            HolidayVariable = c("USPublicHolidays", "EasterGroup", "ChristmasGroup","OtherEcclesticalFeasts"),
            HolidayLags = c(1,2,3),
            HolidayMovingAverages = c(2,3),
            
            # Lags, moving averages, and other rolling stats
            Lags = if(TimeUnit == "weeks") list("weeks" = c(1,2,3,4,5,8,9,12,13,51,52,53), "months" = c(1,2,6,12), "quarters" = c(1,2,3,4)) else if(TimeUnit == "months") list("months" = c(1,2,6,12), "quarters" = c(1,2,3,4)),
            MA_Periods = if(TimeUnit == "weeks") list("weeks" = c(2,3,4,5,8,9,12,13,51,52,53), "months" = c(2,6,12), "quarters" = c(2,3,4)) else if(TimeUnit == "months") list("months" = c(2,6,12), "quarters" = c(2,3,4)),
            # SD_Periods = NULL,
            # Skew_Periods = NULL,
            # Kurt_Periods = NULL,
            # Quantile_Periods = NULL,
            # Quantiles_Selected = NULL,
            
            # Bonus features
            AnomalyDetection = NULL,
            FourierTerms = 0,
            # TimeTrendVariable = TRUE,
            ZeroPadSeries = "maxmax",
              # maxmax - Fill from the absolute min date to the absolute max date
              # minmax - Fill from the max date of the min set to the absolute max date
              # maxmin - Fill from the absolute min date to the min of the max dates
              # minmin - Fill from the max date of the min dates to the min date of the max dates
            DataTruncate = FALSE,
            
            # ML insights
            # PDFOutputPath = NULL,
            # SaveDataPath = NULL,
            # NumOfParDepPlots = 0L,
            
            # ML grid tuning args
            GridTune = TRUE,
            PassInGrid = NULL,
            ModelCount = 5,
            MaxRunsWithoutNewWinner = 50,
            MaxRunMinutes = 5,
            
            # ML loss functions
            EvalMetric = "R2",
            # EvalMetricValue = 1,
            LossFunction = "RMSE",
            # LossFunctionValue = 1,
            
            # ML tuning args
            NTrees = 1500,
            # Depth = Tuning[Run, Depth],
            # Langevin = as.logical(Tuning[Run, Langevin]),
            # DiffusionTemperature = 10000,
            
            # ML overfitting args
            # LearningRate = 0.03,
            # L2_Leaf_Reg = (3.0,4.0),
            # ModelSizeReg = 0.0,
            # RSM = 0.75,
            # MinDataInLeaf = 5,
            # BorderCount = 254,
            # RandomStrength = 0.75,
            # SamplingUnit = "Group",
            
            # ML styles
            # GrowPolicy = ("SymmetricTree","Lossguide","Depthwise"),
            # BootStrapType = c("Bayesian","MVS","No"),
            # FeatureBorderType = "GreedyLogSum",
            # SubSample = NULL,
            ScoreFunction = "Cosine" # if(TaskType == "GPU") "NewtonL2" else "Cosine")
            
            )

**Comment by AdrianAntico** [comment]
The argument TrainOnFull needs to be set to TRUE. When it's set to FALSE the function will build out the regression model and collect evaluation information and return that when complete. When TrainOnFull is TRUE, it will build the model and skip over building out the evaluation information, and create a forecast.

**Comment by joclimber** [comment]
wow! great response time :) 
just ran on TRUE for TrainOnFull and got an error message:

Error in data.table::set(UpdateData., i = N., j = c("Predictions", TargetColumnName.),  : 
  i[1] is 36 which is out of range [1,nrow=35].

![image](https://user-images.githubusercontent.com/85248827/120538296-b56cb380-c3bc-11eb-8d64-c7dc7869c4e9.png)

**Comment by AdrianAntico** [comment]
@joclimber I made a few updates to the code below and it ended up running without issue on my machine. There was a blank space after the column name "month". Also, "value" should be the target variable and "month" should be the date variable. I set them correctly below. One other possibility - if it has been a while since you last installed RemixAutoML, please do so again. I know there were some bug fixes several releases ago, which could be a reason for the issue. 

```
# creating data
library(dplyr)
data <- data.frame(
  month=c('2016-01-01','2016-02-01','2016-03-01','2016-04-01','2016-05-01','2016-06-01','2016-07-01','2016-08-01','2016-09-01','2016-10-01','2016-11-01','2016-12-01','2017-01-01','2017-02-01','2017-03-01','2017-04-01','2017-05-01','2017-06-01','2017-07-01','2017-08-01','2017-09-01','2017-10-01','2017-11-01','2017-12-01','2018-01-01','2018-02-01','2018-03-01','2018-04-01','2018-05-01','2018-06-01','2018-07-01','2018-08-01','2018-09-01','2018-10-01','2018-11-01'),
  value = c(20,280,200,90,100,40,70,210,90,0,10,70,10,90,300,0,70,40,90,100,60,80,40,140,30,130,70,110,60,20,10,220,330,1,134)) %>%
  dplyr::mutate(month = zoo::as.Date(month))

# running function
TimeUnit = "months"

# Run function
CatBoostResults <- RemixAutoML::AutoCatBoostCARMA(

  # data args
  data = data,
  TargetColumnName = "value",
  DateColumnName = "month",
  HierarchGroups = NULL,
  GroupVariables = NULL,
  TimeUnit = TimeUnit,
  TimeGroups = if(TimeUnit == "weeks") c("weeks","months","quarters") else if(TimeUnit == "months") c("months","quarters"),

  # Production args
  TrainOnFull = FALSE,
  PartitionType = "timeseries",
  FC_Periods = 13,
  TaskType = "CPU",
  Timer = TRUE,
  DebugMode = TRUE,

  # Target variable transformations
  TargetTransformation = FALSE,
  Methods = c("BoxCox","Asinh","Log","LogPlus1","YeoJohnson"),
  Difference = FALSE,
  NonNegativePred = TRUE,
  RoundPreds = FALSE,

  # Calendar-related features
  CalendarVariables = if(TimeUnit == "weeks") c("week","wom","month","quarter") else if(TimeUnit == "months") c("month","quarter"),
  HolidayVariable = c("USPublicHolidays", "EasterGroup", "ChristmasGroup","OtherEcclesticalFeasts"),
  HolidayLags = c(1,2,3),
  HolidayMovingAverages = c(2,3),

  # Lags, moving averages, and other rolling stats
  Lags = if(TimeUnit == "weeks") list("weeks" = c(1,2,3,4,5,8,9,12,13,51,52,53), "months" = c(1,2,6,12), "quarters" = c(1,2,3,4)) else if(TimeUnit == "months") list("months" = c(1,2,6,12), "quarters" = c(1,2,3,4)),
  MA_Periods = if(TimeUnit == "weeks") list("weeks" = c(2,3,4,5,8,9,12,13,51,52,53), "months" = c(2,6,12), "quarters" = c(2,3,4)) else if(TimeUnit == "months") list("months" = c(2,6,12), "quarters" = c(2,3,4)),
  # SD_Periods = NULL,
  # Skew_Periods = NULL,
  # Kurt_Periods = NULL,
  # Quantile_Periods = NULL,
  # Quantiles_Selected = NULL,

  # Bonus features
  AnomalyDetection = NULL,
  FourierTerms = 0,
  # TimeTrendVariable = TRUE,
  ZeroPadSeries = "maxmax",
  # maxmax - Fill from the absolute min date to the absolute max date
  # minmax - Fill from the max date of the min set to the absolute max date
  # maxmin - Fill from the absolute min date to the min of the max dates
  # minmin - Fill from the max date of the min dates to the min date of the max dates
  DataTruncate = FALSE,

  # ML insights
  # PDFOutputPath = NULL,
  # SaveDataPath = NULL,
  # NumOfParDepPlots = 0L,

  # ML grid tuning args
  GridTune = TRUE,
  PassInGrid = NULL,
  ModelCount = 5,
  MaxRunsWithoutNewWinner = 50,
  MaxRunMinutes = 5,

  # ML loss functions
  EvalMetric = "R2",
  # EvalMetricValue = 1,
  LossFunction = "RMSE",
  # LossFunctionValue = 1,

  # ML tuning args
  NTrees = 1500,
  # Depth = Tuning[Run, Depth],
  # Langevin = as.logical(Tuning[Run, Langevin]),
  # DiffusionTemperature = 10000,

  # ML overfitting args
  # LearningRate = 0.03,
  # L2_Leaf_Reg = (3.0,4.0),
  # ModelSizeReg = 0.0,
  # RSM = 0.75,
  # MinDataInLeaf = 5,
  # BorderCount = 254,
  # RandomStrength = 0.75,
  # SamplingUnit = "Group",

  # ML styles
  # GrowPolicy = ("SymmetricTree","Lossguide","Depthwise"),
  # BootStrapType = c("Bayesian","MVS","No"),
  # FeatureBorderType = "GreedyLogSum",
  # SubSample = NULL,
  ScoreFunction = "Cosine" # if(TaskType == "GPU") "NewtonL2" else "Cosine")
)
```

**Comment by joclimber** [comment]
Thanks, @AdrianAntico 

Even with those changes applied, I coundn't get past that error. Updated to lastest version, v0.5.3

Do you have any other tips on what can be the solution in this case?

![image](https://user-images.githubusercontent.com/85248827/120545067-8c502100-c3c4-11eb-9ffa-dd7e1d695ca0.png)

**Comment by AdrianAntico** [comment]
@joclimber for the code I pasted back in here, were you able to get that to run correctly?

**Comment by joclimber** [comment]
My mistake, Adrian! I just got it working. Just remapped the variables in the code again and it went smoothly until it finished. Thank you for your responsiveness, patience and great package!

**Comment by AdrianAntico** [comment]
@joclimber thanks! Feel free to let me know how it goes or if you have any feature requests.

**Comment by joclimber** [comment]
Just narrowed down that the error occurred due to variable TimeUnit = "month" instead of TimeUnit = "months"

**Comment by AdrianAntico** [comment]
@joclimber I'm going to expand the list of acceptable names to help prevent that issue. I've done it elsewhere in the package. I'll make that change for the upcoming release.
